File: template/hess-smith/matlab_students/hessSmith.m
% ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ %
%                                                                              %
% Hess-Smith method                                                            %
% for 2-dimensional steady incompressible irrotational flows around an airfoil %
%                                                                              %
%                                                                              %
% ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ %

close all ; clear all ; clc

deg2rad = pi/180 ; % <- this variable should be global (...)


% === Free-stream conditions ===
freeStream.P        = 0.0   ;          % free-stream pressure
freeStream.rho      = 1.225 ;          % free-stream density
freeStream.v        = 10.0  ;          %             velocity absolute value
freeStream.alpha    = 2.0 * deg2rad ;  %             velocity direction
freeStream.vvec     = freeStream.v * [ cos(freeStream.alpha) ; sin(freeStream.alpha) ] ;

% === Airfoil input ===
% Airfoil structure definition
airfoil.id           = 1 ;
airfoil.airfoil_str  = 'NACA0012' ;
airfoil.chord        = 1.0 ;
airfoil.theta        = 0.0 * deg2rad ;
airfoil.xcRefPoint   = 0.25 ;
airfoil.refPoint     = [ 0.0 ; 0.0 ] ;
airfoil.nChordPanels = 30  ;

% === Build Geometry ===
% OBS: all the airfoils have TE -> nAirfoils = nTE = n. Kutta conditions required
% 1. Build the required arrays and structures
[ rr , ee , ii_te , elems , nelems , npoints ] = build_geometry( airfoil ) ;

% % 2. Fill some matrices with the elems info, because MATLAB is meant to efficiently
% % work with matrices
% rrc  = zeros(2,nelems) ;
% tvers= zeros(2,nelems) ;
% nvers= zeros(2,nelems) ;
% for ie = 1 : nelems
%    rrc(  :,ie) = elems(ie).cen ;
%    tvers(:,ie) = elems(ie).tver ;
%    nvers(:,ie) = elems(ie).nver ;
% end

% % === Build the linear system ===
% [ A , b , Au , Av ] = build_linsys( freeStream , elems , ii_te )  ;
% 
% % === Solve the linear system ===
% % vector vv contains the intensity of the singularities used to model solid bodies
% vv = A \ b;
% 
% % === Retrieve physical quantities === 
% % x, y components of the velocity field on the body ----------------------------
% u = Au * vv + freeStream.v*cos(freeStream.alpha);
% v = Av * vv + freeStream.v*sin(freeStream.alpha);
% uvec = [ u , v ] ;
% % modu = sqrt(u.^2 + v.^2);
% % angolo = atan2(v,u);
% 
% % Tangent and normal components of the velocity field w.r.t. the surface -------
% % !!! check that vNi is = 0 (approximately)                                  !!!
% vTi =  sum( uvec' .* tvers ) ;
% vNi =  sum( uvec' .* nvers ) ;
% 
% % Pressure coefficient ---------------------------------------------------------
% % !!! check that cP is approximately equal to 1.0 at the stagnation point    !!!
% % !!! close to the leading edge                                              !!!
% cP = 1 - vTi.^2 ./freeStream.v^2;
% 
% % Plot cP -----
% figure ; hold on
% plot(rr(1,:),rr(2,:),'-') 
% plot(rrc(1,:),cP,'-o')
% grid on, axis equal, hold off

File: template/hess-smith/python_students/hessSmith.py
# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ #
#                                                                              #
# Hess-Smith method                                                            #
# for 2-dimensional steady incompressible irrotational flows around an airfoil #
#                                                                              #
#                                                                              #
# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ #
#                                                                              #
# The implementation exploits python dictionaries (DICT) and lists (LIST)      #
# They make the implementation easier to understand (or they should, at least) #
#                                                                              #
# Some suggestions for newbies:                                                #
# - python counts starting from 0, in an awkward way. Some examples:           #
#  a[0] is the first elem of the array a                                       #
#  a[0:4] are the first 4 (!!!!) elems of the array a                          #
#  if the index of the array exceed the dimension of the array, the count      #
#   restarts from zero, without any error (some "circular" indices)            #
# - lots of math operations come with numpy. To use them, you need to call     #
#  them with numpy (or its alias, e.g. import numpy as np, np is the alias).   #
#  E.g. numpy.cos(theta) for the cos() function, or np.cos() if the alias is   #
#  defined                                                                     #
# - Dictionaries are 'indexed' through keys. If a dictionary is defined as     #
#  dict_example = { 'name':'paolino', 'surname':'paperino', 'age':100, \       #
#                   'height':1.20 }                                            #
#  you can access its contents, using its KEYs. If you want to know the value  #
#  or the height field,                                                        #
#                                                                              #
#  paperino_height = dict_example['height']                                    #
#                                                                              #
# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ #

import numpy as np                  # py library for math
import matplotlib.pyplot as plt     # py library for plot

import build_geometry as geo        # file with the function that builds the geom

deg2rad = np.pi/180.

#> === Free-stream conditions ===
#> Set Reynolds number of the analysis, Re1 = rho * U * c1 / mu, 
# for airfoils w/ unitary chord c1 = 1.:
# -> infinite value sets inviscid analysis
# -> finite values  turn on viscous corrections
Re1 = np.inf

#> Initialize freeStream DICT with input data for an inviscid analysis
freeStream = { 'p': 0., 'rho': 1., 'U': 1., 'al': 2.*deg2rad, \
               'viscous_correction':False }

#> Update freeStream DICT, with some useful quantities:
# free stream velocity vector
freeStream['Uvec']:np.array([ np.cos(freeStream['al']) , \
                              np.sin(freeStream['al'])]) *freeStream['U']

# viscosity, if viscous corrections are required (with finite value of Re1)
if ( Re1 != np.inf ):
  freeStream['viscous_corrections']:True
  freeStream['Re1']:Re1
  freeStream['dyn_visc']:freeStream['rho']*freeStream['U']/Re1
  freeStream['kin_visc']:freeStream['dyn_visc']/freeStream['rho']

#> === Airfoil input ===
#> Initialize bodies DICT, containing the list of the DICTs of the bodies
# of the simulation: in our lab, we will use just one body, whose name
# (KEY) is 'Airfoil_1'. At the initialization its value is void, []
bodies = { 'Airfoil_1':[] }
#> Initialize bodies['Airfoil_1'] DICT with airfoil data
bodies['Airfoil_1'] = { 'id':1, 'airfoil':'NACA0012', 'chord':1., \
                      'theta': 0.*deg2rad, 'xc_ref_point':0.25, \
                      'ref_point':np.array([ 0., 0. ]), \
                      'n_chord_pan':15 }

#> === Define the geometry of the body ( see build_geometry.py ) ===
elems, rr, ee, ii_te = geo.build_geometry( bodies['Airfoil_1'] )

#> Check output of build_geometry()
print(' np.shape(elems): '); print(np.shape(elems))
print(' np.shape(rr)   : '); print(np.shape(rr))
print(' np.shape(ee)   : '); print(np.shape(ee))
print(' np.shape(ee_te): '); print(np.shape(ee_te))
# print(' elems: ')
# for i in np.arange(len(elems)):
#   print(i, ': ', elems[i])
# print(' rr: '); print(rr)
# print(' ee: '); print(ee)

#> Check geometry
plt.plot(rr[0,:], rr[1,:], '-o')
for i in np.arange( np.size(elems,0) ):
  plt.quiver( elems[i]['cen'][0], elems[i]['cen'][1], \
              elems[i]['nor'][0], elems[i]['nor'][1], \
              color='black', width=0.005, headwidth=3 )
  plt.quiver( elems[i]['cen'][0], elems[i]['cen'][1], \
              elems[i]['tan'][0], elems[i]['tan'][1], \
              color='red', width=0.005, headwidth=3 )
plt.axis('equal')



#> === Build the linear system ===
# A, b, Au, Av = linsys.build_linsys( freeStream, elems, ii_te )

#> === Solve linear system ===
# ...

#> === Retrieve physical quantities ===
#> compontents of the velocity
# u = ...
# v = ...

#> pressure, pressure coefficients 
# ...

#> === Integral loads ===
#> aerodynamic coefficients
# ...


#> === Viscous correction ===
# ...













plt.show()
